### ISSUE TITLE
ParameterName: new option to skip methods Override annotation

### ISSUE THREAD

**Issue by romani** [checkstyle/checkstyle#2290]
base on discussion in #985 .

1) we need an option for http://checkstyle.sourceforge.net/config_naming.html#ParameterName
to skip violations on parameter that are for methods with annotation ```@Override```.

reason:
```
.gui.TreeTableCellRenderer:100  Avoid variables with short names like h.
.gui.TreeTableCellRenderer:100  Avoid variables with short names like w.
.gui.TreeTableCellRenderer:100  Avoid variables with short names like x.
.gui.TreeTableCellRenderer:100  Avoid variables with short names like y.
```

2) We need to update default usage of ```<module name="ParameterName"/>``` at 
https://github.com/checkstyle/checkstyle/blob/master/config/checkstyle_checks.xml#L180
to custom config to enforce name to be 2 or more symbols in size. Format "^[a-z][a-zA-Z0-9]+$".

3) Please refactor this Check and make option "boolean skipCatchParameter = true". TRUE by default to keep compatibility with previous behavior.

Reason: right not our config allow one char variables.

**Comment by pbaranchikov** [comment]
Hi, @romani 
>1) we need an option for http://checkstyle.sourceforge.net/config_naming.html#LocalVariableName
to skip violations on parameter that are for methods with annotation @Override.

I still cannot understand, why do we need this change. If I override the method with malformed parameter names, I can rename formal parameters in a way that they conform my LocalVariableName requirements.

**Comment by romani** [comment]
@pbaranchikov , 

1) When user "Override" , he kind of position to follow style of method author, it is better to not change signature. Most of engineers just follow IDEs auto-generation and that is right way. We just need to let Checkstyle easily ignore cases when class/method is coming from outside(thirdparty library with different style) and we can not fix this, but we have to use it.

2) please propose names for TreeTableCellRenderer:100 .

**Comment by pbaranchikov** [comment]
We can face the following problem
```java
public class A {
   public void foo(String barVar);
}
public class B extnds A {
    @Override
    public void foo(String BAR_VAR);
}
```
This should not be restricted. WIth the option, you will be free to name method parameters as you want, despite of any styling rules in overridden methods.

If you insist on adding this option, I vote to make it disabled by default.

>2) please propose names for TreeTableCellRenderer:100 .

- ttCr
- treeTableCR
- ttCellRenderer
...

**Comment by romani** [comment]
> 1)

:) , default should work for most users. Can you show me examples in opensource where people change name of arguments from original ?
If no example found but If you are in a team where such code(from your example) is likely to happen, you need to use specially configured Check. Very strict defaults force user to disable Check completely without considering reading documentation for Check.

> 2)
I meant variables at 
https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/gui/TreeTableCellRenderer.java#L100 . Names for "x" , "y" .

**Comment by mkordas** [comment]
@pbaranchikov 
If you want to prevent situation when parent parameter is `barVar` while overridden one is `BAR_VAR` you should use other check like IntelliJ's `Parameter name differs from parameter in overridden method`.

**Comment by MEZk** [comment]
@romani 
Right now the Check does not take into account method parameters at all.

**Example**

Test.java:
````java
public class InputOverrideAnnotation {

    @Override
    public boolean equals(Object o) { // there should be a violation
        return super.equals(o);
    }
}
````

Config (my_check.xml):
````xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">


<module name = "Checker">
    <property name="charset" value="UTF-8"/>
    <property name="severity" value="warning"/>
    
    <module name="TreeWalker">
         <module name="LocalVariableName">
	    <property name="format" value="^ABC$"/>
	    <property name="tokens" value="PARAMETER_DEF"/>
	</module>
    </module>
</module>
````

Checkstyle's  6.12 RELEASE version output:
```
java -jar checkstyle-6.12-all.jar -c my_check.xml Test.java
Starting audit...
Audit done.
```

Problem is caused by [ScopeUtils.isLocalVariableDef(DetailAST aAST)](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/utils/ScopeUtils.java#L251) method which is used in [mustCheckName(DetailAST ast)](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalVariableNameCheck.java#L127) method of LocalVariableNameCheck.

Take a look at isLocalVariableDef. When [ast.getType is PARAMETER_DEF](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/utils/ScopeUtils.java#L262) it checks only for LITERAL_CATCH token type of parent. But when we parse method the type of the parent of PARAMETER_DEF token is PARAMETERS. That is why mustCheckName return false for method parameters tokens.

In addition to this, testCatchParameter test in LocalVariableNameCheckTest must not pass, because in input file 'InputEmptyStatement' the parameter of EmptyStatements method does not match the ```^e$``` regexp patter which is specified in the test.

**Comment by MEZk** [comment]
@romani 
To save compatibility with other checks I propose to implement fix directly in LocalVariableName and do not change ScopeUtils.isLocalVariableDef .

**Comment by romani** [comment]
yes, we have special Check for this - http://checkstyle.sourceforge.net/config_naming.html#ParameterName, let me review whole issue.

**Comment by MEZk** [comment]
@romani 
I guess, that PARAMETER_DEF token is used only for catch blocks, not for the methods.

**Comment by romani** [comment]
@MEZk , please recheck why ParameterName is not reporting such names in our code. I do not see override annotation in its code, default value of check is also does not looks like allow one character names.

**Comment by MEZk** [comment]
@romani 
For ParameterName format is ^[a-z][a-zA-Z0-9]*$ by default.
Quantifier  * means zero or more times. That is why the Check allows single character names.

Proof: https://regex101.com/
![screenshot-1](https://cloud.githubusercontent.com/assets/7242568/10941915/a8c519cc-831e-11e5-9aef-b37f49c56d9e.png)
![screenshot](https://cloud.githubusercontent.com/assets/7242568/10941914/a8be2414-831e-11e5-8e0c-f5e38dfd0d3c.png)

**Comment by romani** [comment]
@MEZk ,

> 2) We need to update default usage of <module name="ParameterName"/> at
https://github.com/checkstyle/checkstyle/blob/master/config/checkstyle_checks.xml#L180
to custom config to enforce name to be 2 or more symbols in size. Format "^[a-z][a-zA-Z0-9]+$".

not completed yet.

> 3) Please refactor this Check and make option "boolean skipCatchParameter = true". TRUE by default to keep compatibility with previous behavior.

not completed yet.

**Comment by romani** [comment]
merged.
